File: knotlink-sdk-qt/knotlink/tcpclient.cpp
/*
 * KnotLink SDK - C++
 * Copyright (c) 2024-2026 KnotLink Contributors
 * SPDX-License-Identifier: MIT
 */

#include "tcpclient.h"
#include <QDebug>
#include <QDataStream>
#include <QtEndian>

Q_LOGGING_CATEGORY(knotlinkTcp, "knotlink.tcp")

const quint32 MAX_MSG_SIZE = 16 * 1024 * 1024; // 16MB，与服务器一致

TcpClient::TcpClient(QObject *parent) : QObject(parent) {
	tcpSocket = new QTcpSocket(this);
	heartBeatTimer = new QTimer(this);
	heartBeatTimer->setInterval(180000);  // 3分钟

	// 信号在构造函数中连接，避免 connectToServer 后错过事件
	connect(heartBeatTimer, &QTimer::timeout, this, &TcpClient::sendHeartbeat);
	connect(tcpSocket, &QTcpSocket::connected, this, &TcpClient::socketConnected);
	connect(tcpSocket, &QTcpSocket::disconnected, this, &TcpClient::socketDisconnected);
	connect(tcpSocket, &QTcpSocket::readyRead, this, &TcpClient::readData);
}

// Qt parent 机制自动管理子对象生命周期，无需手动 delete

void TcpClient::connectToServer(const QString &ip, uint16_t port) {
	tcpSocket->connectToHost(QHostAddress(ip), port);

	if (!tcpSocket->waitForConnected(3000)) {
		qCWarning(knotlinkTcp) << "Connection failed:" << tcpSocket->errorString();
		return;
	}

	heartBeatTimer->start();
}

void TcpClient::socketConnected() {
	sendHeartbeat();
	emit connected();
}

void TcpClient::socketDisconnected() {
	heartBeatTimer->stop();
	emit disconnected();
}

// ------------------------------------------------------------
// 写入数据：添加 4 字节大端长度前缀
// ------------------------------------------------------------
void TcpClient::writeWithLengthPrefix(const QByteArray &data) {
	QByteArray block;
	QDataStream out(&block, QIODevice::WriteOnly);
	out.setVersion(QDataStream::Qt_5_0);
	out << (quint32)data.size();   // 写入长度（4 字节，大端）
	block.append(data);            // 追加消息体
	tcpSocket->write(block);
}

// 对外发送数据（自动加长度前缀）
void TcpClient::sendData(const QByteArray &data) {
	if (tcpSocket->state() == QAbstractSocket::ConnectedState) {
		writeWithLengthPrefix(data);
	} else {
		qCWarning(knotlinkTcp) << "Cannot send data, disconnected";
	}
}

// ------------------------------------------------------------
// 接收数据：追加到缓冲区，然后解析完整消息
// ------------------------------------------------------------
void TcpClient::readData() {
	QByteArray newData = tcpSocket->readAll();
	if (newData.isEmpty()) return;
	buffer.append(newData);
	processBuffer();
}

// ------------------------------------------------------------
// 解析缓冲区：根据长度前缀提取完整消息
// ------------------------------------------------------------
void TcpClient::processBuffer() {
	while (true) {
		if (buffer.size() < 4) break; // 长度字段未完整
		
		// 读取长度（大端序）
		quint32 len = qFromBigEndian<quint32>((const uchar*)buffer.constData());
		
		if (len == 0 || len > MAX_MSG_SIZE) {
			qCCritical(knotlinkTcp) << "Invalid message length:" << len << ", disconnecting";
			tcpSocket->disconnectFromHost();
			return;
		}
		
		if (buffer.size() < (int)(len + 4)) break; // 消息体未完整
		
		// 提取消息体（跳过长度字段）
		QByteArray msg = buffer.mid(4, len);
		buffer.remove(0, len + 4);
		
		// 处理消息：忽略心跳响应，其他发出信号
		if (msg == "heartbeat_response") {
			qCDebug(knotlinkTcp) << "Heartbeat response received, ignoring";
		} else {
			emit receivedData(msg);
		}
		// 继续循环，可能还有更多完整消息
	}
}

// ------------------------------------------------------------
// 发送心跳：使用带长度前缀的格式
// ------------------------------------------------------------
void TcpClient::sendHeartbeat() {
	if (tcpSocket->state() == QAbstractSocket::ConnectedState) {
		writeWithLengthPrefix("heartbeat");
		qCDebug(knotlinkTcp) << "Heartbeat sent";
	} else {
		qCDebug(knotlinkTcp) << "Cannot send heartbeat, disconnected";
	}
}

// 错误处理（可选）
void TcpClient::handleError(QAbstractSocket::SocketError socketError) {
	qCWarning(knotlinkTcp) << "Socket error:" << socketError << "-" << tcpSocket->errorString();
}

File: knotlink-sdk-pyqt/knotlink-pyqt/knotlink-pyqt/tcpclient.py
# KnotLink SDK - Python
# Copyright (c) 2024-2026 KnotLink Contributors
# SPDX-License-Identifier: MIT

from PyQt5.QtCore import QObject, pyqtSignal, QTimer
from PyQt5.QtNetwork import QTcpSocket, QHostAddress, QAbstractSocket
import struct

class TcpClient(QObject):
    connected = pyqtSignal()
    disconnected = pyqtSignal()
    receivedData = pyqtSignal(bytes)  # 发出完整消息（不含长度前缀）

    MAGIC = b'KK\x00\x02'  # 协议魔数：KK + 版本号 2.0

    def __init__(self, parent=None):
        super().__init__(parent)
        self.tcpSocket = QTcpSocket(self)
        self.heartBeatTimer = QTimer(self)
        self.heartBeatTimer.setInterval(180000)  # 3分钟
        self.buffer = bytearray()                 # 接收缓冲区
        self.MAX_MSG_SIZE = 16 * 1024 * 1024      # 16MB

        self.tcpSocket.connected.connect(self.socketConnected)
        self.tcpSocket.disconnected.connect(self.socketDisconnected)
        self.tcpSocket.readyRead.connect(self.readData)
        self.tcpSocket.errorOccurred.connect(self.handleError)
        self.heartBeatTimer.timeout.connect(self.sendHeartbeat)

    # ---------- 连接 ----------
    def connectToServer(self, ip, port):
        self.tcpSocket.connectToHost(QHostAddress(ip), port)
        if not self.tcpSocket.waitForConnected(3000):
            print("连接失败：", self.tcpSocket.errorString())
            return False
        self.heartBeatTimer.start()
        return True

    def socketConnected(self):
        self.connected.emit()

    def socketDisconnected(self):
        self.disconnected.emit()
        self.heartBeatTimer.stop()

    # ---------- 发送数据（自动加长度前缀） ----------
    def _writeWithLengthPrefix(self, data: bytes):
        if self.tcpSocket.state() != QAbstractSocket.ConnectedState:
            print("无法发送数据，连接已断开。")
            return False
        if len(data) > self.MAX_MSG_SIZE:
            print(f"消息过长: {len(data)} > {self.MAX_MSG_SIZE}")
            return False
        # 构造 4 字节大端长度前缀 + 魔数前缀
        length_bytes = struct.pack('>I', len(data))
        # 发送魔数 + 长度 + 数据
        self.tcpSocket.write(self.MAGIC + length_bytes + data)
        return True

    def sendData(self, data: bytes):
        self._writeWithLengthPrefix(data)

    # ---------- 心跳 ----------
    def sendHeartbeat(self):
        if self.tcpSocket.state() == QAbstractSocket.ConnectedState:
            if self._writeWithLengthPrefix(b"heartbeat"):
                print("发送心跳包成功")
        else:
            print("无法发送心跳包，连接已断开。")
            self.heartBeatTimer.stop()

    # ---------- 接收数据 ----------
    def readData(self):
        # 读取所有可用数据
        new_data = self.tcpSocket.readAll().data()
        if not new_data:
            return
        self.buffer.extend(new_data)
        self._processBuffer()

    def _processBuffer(self):
        """从缓冲区解析完整消息"""
        MAGIC_LEN = len(self.MAGIC)  # 4
        HEADER_LEN = MAGIC_LEN + 4   # 8
        while True:
            if len(self.buffer) < HEADER_LEN:
                break  # 协议头未完整

            # 校验魔数
            if self.buffer[:MAGIC_LEN] != self.MAGIC:
                print(f"魔数不匹配: expected {self.MAGIC.hex()}, got {bytes(self.buffer[:MAGIC_LEN]).hex()}，断开连接")
                self.tcpSocket.disconnectFromHost()
                return

            # 读取长度（大端，在魔数之后）
            length = struct.unpack('>I', self.buffer[MAGIC_LEN:HEADER_LEN])[0]
            if length == 0 or length > self.MAX_MSG_SIZE:
                print(f"无效消息长度: {length}，断开连接")
                self.tcpSocket.disconnectFromHost()
                return

            if len(self.buffer) < HEADER_LEN + length:
                break  # 消息体未完整

            # 提取消息体（跳过魔数 + 长度）
            msg = bytes(self.buffer[HEADER_LEN:HEADER_LEN+length])
            del self.buffer[:HEADER_LEN+length]

            # 处理消息：忽略心跳响应
            if msg == b"heartbeat_response":
                print("收到心跳响应，忽略")
            else:
                print("收到数据：", msg)
                self.receivedData.emit(msg)
            # 继续循环处理剩余缓冲区

    # ---------- 错误处理 ----------
    def handleError(self, socketError):
        print("错误：", socketError, "-", self.tcpSocket.errorString())

    # ---------- 断开 ----------
    def disconnect(self):
        self.tcpSocket.disconnectFromHost()
        self.buffer.clear()